File: leetcode/algorithms/101_symmetric_tree/main.cs
using System.Collections.Generic;

public class SymmetricTree {
    public class TreeNode {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public bool IsSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }

        return CheckNode(root.left, root.right);
    }

    public bool CheckNode(TreeNode left, TreeNode right) {
        if (left == null && right == null) {
            return true;
        }

        if (left == null || right == null || left.val != right.val) {
            return false;
        }

        return CheckNode(left.left, right.right) && CheckNode(left.right, right.left);
    }


    // Solution
    // Solution 1: Recursive
    public bool solution1(TreeNode root) => solution1_CheckSymetry(root?.left, root?.right);

    private bool solution1_CheckSymetry(TreeNode left, TreeNode right) {
        if (left == null || right == null) {
            return left?.val == right?.val;
        }

        if (left.val != right.val) {
            return false;
        }

        return solution1_CheckSymetry(left.left, right.right) && solution1_CheckSymetry(left.right, right.left);
    }

    // Solution 2: Iterative DFS
    public bool solution2(TreeNode root) {
        if (root == null) {
            return true;
        }

        var stack = new Stack<(TreeNode, TreeNode)>();
        stack.Push((root.left, root.right));

        while (stack.Any()) {
            switch (stack.Pop()) {
                case (null, null):
                    continue;
                case (null, _) or (_, null):
                    return false;
                case (TreeNode l, TreeNode r) when l.val != r.val:
                    return false;
                case (TreeNode l, TreeNode r) when l.val == r.val:
                    stack.Push((l.left, r.right));
                    stack.Push((l.right, r.left));
                    break;
            }
        }

        return true;
    }
}

File: leetcode/algorithms/101_symmetric_tree/main.dart
import 'dart:collection';

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

class SymmetricTree {
  /**
   * Recursion - DFS
   *
   * Complexities:
   *   N - The number of nodes in `root`
   *   H - The heights of `root`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(H)
   */
  bool isSymmetric(TreeNode? root) {
    if (root == null) {
      return true;
    }

    bool checkSymmetric(TreeNode? p, TreeNode? q) {
      if (p == null && q == null) {
        return true;
      }

      if (p != null && q != null) {
        return checkSymmetric(p.left, q.right) &&
            checkSymmetric(p.right, q.left) &&
            p.val == q.val;
      }

      return false;
    }

    return checkSymmetric(root.left, root.right);
  }


  // Solution
  /**
   * Solution 1
   *
   * Recursion - DFS
   *
   * Complexities:
   *   N - The number of nodes in `root`
   *   H - The heights of `root`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(H)
   */
  bool solution1(TreeNode? root) {
    if (root == null) {
      return true;
    }

    return _isMirror(root.left, root.right);
  }

  bool _isMirror(TreeNode? t1, TreeNode? t2) {
    if (t1 == null && t2 == null) {
      return true;
    }

    if (t1 == null || t2 == null) {
      return false;
    }

    if (t1.val != t2.val) {
      return false;
    }

    return _isMirror(t1.left, t2.right) && _isMirror(t1.right, t2.left);
  }

  /**
   * Solution 2
   *
   * Iteration
   *
   * Complexities:
   *   N - The number of nodes in `root`
   *   H - The heights of `root`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(H)
   */
  bool solution2(TreeNode? root) {
    if (root == null) {
      return true;
    }

    Queue<TreeNode?> queue = Queue();

    queue.add(root.left);
    queue.add(root.right);

    while (queue.isNotEmpty) {
      TreeNode? t1 = queue.removeFirst();
      TreeNode? t2 = queue.removeFirst();

      if (t1 == null && t2 == null) {
        continue;
      }

      if (t1 == null || t2 == null || t1.val != t2.val) {
        return false;
      }

      queue.add(t1.left);
      queue.add(t2.right);

      queue.add(t1.right);
      queue.add(t2.left);
    }

    return true;
  }
}
